cpp/docs/TESTING.md

```diff
@@ -0,0 +1,408 @@
+// [ {{0, 1}, NULL}, {{4, 5}, NULL} ]
+auto validity = make_counting_transform_iterator(0, [](auto i){return i%2;});
+lists_column_wrapper l{ {{{0, 1}, {2, 3}}, validity}, {{{4, 5}, {6, 7}}, validity} };
+```
+
+#### `structs_column_wrapper`
+
+The `structs_column_wrapper` class should be used to create columns of structs. It provides 
+constructors that accept a vector or initializer list of pre-constructed columns or column wrappers
+for child columns. For nullable columns, an additional iterator can be provided to indicate the 
+validity of each struct.
+
+Examples:
+```c++
+// The following constructs a column for struct< int, string >.
+auto child_int_col = fixed_width_column_wrapper<int32_t>{ 1, 2, 3, 4, 5 }.release();
+auto child_string_col = string_column_wrapper {"All", "the", "leaves", "are", "brown"}.release();
+
+std::vector<std::unique_ptr<column>> child_columns;
+child_columns.push_back(std::move(child_int_col));
+child_columns.push_back(std::move(child_string_col));
+
+struct_column_wrapper struct_column_wrapper{
+  child_cols,
+  {1,0,1,0,1} // Validity
+};
+
+auto struct_col {struct_column_wrapper.release()};
+```
+
+```c++
+// The following constructs a column for struct< int, string >.
+fixed_width_column_wrapper<int32_t> child_int_col_wrapper{ 1, 2, 3, 4, 5 };
+string_column_wrapper child_string_col_wrapper {"All", "the", "leaves", "are", "brown"};
+
+struct_column_wrapper struct_column_wrapper{
+  {child_int_col_wrapper, child_string_col_wrapper}
+  {1,0,1,0,1} // Validity
+};
+
+auto struct_col {struct_column_wrapper.release()};
+```
+
+```c++
+// The following constructs a column for struct< int, string >.
+fixed_width_column_wrapper<int32_t> child_int_col_wrapper{ 1, 2, 3, 4, 5 };
+string_column_wrapper child_string_col_wrapper {"All", "the", "leaves", "are", "brown"};
+
+struct_column_wrapper struct_column_wrapper{
+  {child_int_col_wrapper, child_string_col_wrapper}
+  cudf::test::make_counting_transform_iterator(0, [](auto i){ return i%2; }) // Validity
+};
+
+auto struct_col {struct_column_wrapper.release()};
+```
+
+### Column Comparison Utilities
+
+A common operation in testing is verifying that two columns are equal, or equivalent, or that they
+have the same metadata.
+
+#### `expect_column_properties_equal`
+
+Verifies that two columns have the same type, size, and nullability. For nested types, recursively 
+verifies the equality of type, size and nullability of all nested children.
+
+#### `expect_column_properties_equivalent`
+
+Verifies that two columns have equivalent type and equal size, ignoring nullability. For nested 
+types, recursively verifies the equivalence of type, and equality of size of all nested children,
+ignoring nullability.
+
+Note "equivalent type". Most types are equivalent if and only they are equal. `fixed_point` types
+are the exception. They are equivalent if the representation type is equal, even if they have 
+different scales.
```

Comment:
One additional point here:

- Nested type columns can be equivalent in the case where they are both of size 0, but one of them has children (also empty) and the other does not.  For cases where size is > 0, both equals and equivalent expect the same # of children.